api/envoy/config/common/matcher/v3/matcher.proto

```diff
@@ -16,6 +18,112 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
 // [#protodoc-title: Unified Matcher API]
 
+// A matcher, which may traverse a matching tree in order to result in a match action.
+message Matcher {
+  // What to do if a match is successful.
+  message OnMatch {
+    oneof on_match {
+      // Nested matcher to evaluate.
+      // If the nested matcher does not match and does not specify
+      // on_no_match, then this matcher is considered not to have
+      // matched, even if a predicate at this level or above returned
+      // true.
+      Matcher matcher = 1;
+
+      // Protocol-specific action to take.
+      core.v3.TypedExtensionConfig action = 2;
+    }
+  }
+
+  // A linear list of field matchers.
+  // The field matchers are evaluated in order, and the first match
+  // wins.
+  message MatcherList {
+    // Predicate to determine if a match is successful.
+    message Predicate {
+      // Predicate for a single input field.
+      message SinglePredicate {
+        // Protocol-specific specification of input field to match on.
+        core.v3.TypedExtensionConfig input = 1;
+
+        oneof matcher {
+          // Use existing infrastructure for actually matching the
+          // value.
+          type.matcher.v3.ValueMatcher value_match = 2;
+
+          // Extension for custom matching logic.
+          core.v3.TypedExtensionConfig custom_match = 3;
+        }
+      }
+
+      // A list of matchers. Used to allow using a list within a oneof.
+      message PredicateList {
+        repeated Predicate predicate = 1;
+      }
+
+      oneof match_type {
+        // A single predicate to evaluate.
+        SinglePredicate single_predicate = 1;
+
+        // A list of predicates to be OR-ed together.
+        PredicateList or = 2;
+
+        // A list of predicates to be AND-ed together.
+        PredicateList and = 3;
+      }
+    }
+
+    // An individual matcher.
+    message FieldMatcher {
+      Predicate predicate = 1;
+      // Determines if the match succeeds.
+
+      OnMatch on_match = 2;
+      // What to do if the match succeeds.
+    }
+
+    // A list of matchers. First match wins.
+    repeated FieldMatcher matchers = 1;
+  }
+
+  // A map of configured matchers. Used to allow using a map within a oneof.
+  message MatcherTree {
+    message MatchMap {
+      map<string, OnMatch> map = 1;
+    }
+
+    // Protocol-specific specification of input field to match on.
+    core.v3.TypedExtensionConfig input = 1;
+
+    // Exact or prefix match maps in which to look up the input value.
+    // If the lookup succeeds, the match is considered successful, and
+    // the corresponding OnMatch is used.
+    oneof tree_type {
+      MatchMap exact_match_map = 2;
+
+      // Longest matching prefix wins.
+      MatchMap prefix_match_map = 3;
```

Comment:
@htuch @markdroth One thing we didn't talk about here was whether we want to keep these proto maps or go for a list as @mattklein123 suggested. There's also a general question on whether we want this built in matcher to be able to mix and match exact and prefix (which a trie easily allows as long we're clear about resolving ties).

List approach could look like:

```
message TreeMatch {
  oneof match_type {
    string exact_match = 1;
    string prefix_match = 2;
  }
}

repeated TreeMatcher matchers = 1;
```

or we could mix and match by adding additional value to the map value:

```
message TreeMatch {
  OnMatch on_match = 1;
  enum MatchType {
    EXACT;
    PREFIX;
  }
  MatchType match_type = 2;
}

map<string, TreeMatch> match_map = 1;
```

I wouldn't expect mixing and matching exact and prefix would be that useful (and the no_match recursive tree lets you try exact matching and then falling back to prefix matching), but it would allow us to optimize that case into a single trie lookup.

Reply:
I think I actually prefer the map approach, since the list approach basically winds up being the same as the `MatchList` approach.  The whole point of having a different representation for sublinear matchers is that it should be obvious from the config that there's something different going on in the implementation.

Reply:
I've seen different options on whether people like reading maps or lists. Personally, I find maps represented in YAML to be confusing and I would prefer a list based API, but I don't feel that strongly about it.

I don't have a strong preference on mixing and matching though I would generally try to keep it as simple as possible so unless there is a strong argument I would skip it.

Reply:
Ok since there's no strong objection I'll leave it as this. 

If we end up wanting a mixed matcher we could always add one later or add an extension that does this